File: leetcode/0020-valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = list()
        
        i = 0
        while i < len(s):
            if s[i] == '(':
                stack.append(')')
            elif s[i] == '{':
                stack.append('}')
            elif s[i] == '[':
                stack.append(']')
            elif len(stack) == 0 or stack[-1] != s[i]:
                return False
            else:
                stack.pop()
            
            i += 1
        
        return len(stack) == 0

File: leetcode/0020-valid-parentheses/solution.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stac;
        int i = 0;
        while (i < s.size()) {
            if (s[i] == '(') {
                stac.push(')');
            } else if (s[i] == '{') {
                stac.push('}');
            } else if (s[i] == '[') {
                stac.push(']');
            } else if (stac.empty() || stac.top() != s[i]) {
                return false;
            } else {
                stac.pop();
            }
            
            i++;
        }
        
        return stac.empty();
    }
};